File: longestSubstringWithReplacement.js
class Solution {
  /**
   * @param {string} s
   * @param {number} k
   * @return {number}
   */
  characterReplacement(s, k) {
    let res = 0;
    let maxF = 0;
    let l = 0;
    let seen = {};

    for (let r = 0; r < s.length; r++) {
      seen[s[r]] = (seen[s[r]] || 0) + 1;
      maxF = Math.max(maxF, seen[s[r]]);
      while (r - l + 1 - maxF > k) {
        seen[s[l]]--;
        l++;
      }
      res = Math.max(res, r - l + 1);
    }

    return res;
  }
}

File: longest_substring_with_replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        res = 0
        max_f = 0
        l = 0
        seen = {}

        for r in range(len(s)):
            seen[s[r]] = 1 + seen.get(s[r], 0)
            max_f = max(max_f, seen[s[r]])
            while (r - l) + 1 - max_f > k:
                seen[s[l]] -= 1
                l += 1

            res = max(res, (r - l) + 1)

        return res
